lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp

```diff
@@ -401,146 +575,231 @@ void AccessConflictAnalysis::identifyBeginAccesses() {
+    return;
 
-    const DenseAccessSet &predSet = mapI->second;
-    mergedAccesses.merge(SparseAccessSet(predSet, result));
-  }
-}
+  auto *endStorage = getRegionStorageFromIns(endAccess);
+  auto &inScope = endStorage->getInScopeAccesses();
+  if (std::find(inScope.begin(), inScope.end(), beginAccess) == inScope.end())
+    return;
 
-// Compute access reachability within the given block.
-void AccessConflictAnalysis::visitBlock(SILBasicBlock *BB) {
-  // Sparse set for tracking accesses with an individual block.
-  SparseAccessSet accessSet(result);
-  mergePredAccesses(BB, accessSet);
+  LLVM_DEBUG(llvm::dbgs() << "No conflict on one path from " << *beginAccess
+                          << " to " << *endAccess);
+  endStorage->removeInScopeAccess(beginAccess);
+  LLVM_DEBUG(llvm::dbgs() << "Got out of scope from " << *beginAccess << " to "
+                          << *endAccess
+                          << " with no conflicts seen (so far)\n");
+  endStorage->addOutOfScopeAccess(beginAccess);
+}
 
-  for (auto &I : *BB) {
+void RegionStorageAnalysis::processBlockRegion(unsigned id) {
+  assert(regionToStorageMap.find(id) == regionToStorageMap.end() &&
+         "Should not process a region twice");
+  auto *region = LRFI->getRegion(id);
+  assert(region->isBlock() && "Expected a block");
+  // A block's inital set is all accesses
+  auto *storage = new RegionStorage(region, result, storageAccesses);
+  regionToStorageMap.insert(std::make_pair(id, storage));
+  mergeBBPredAccesses(id);
+  auto *bb = region->getBlock();
+  for (auto &I : *bb) {
     if (auto *BAI = dyn_cast<BeginAccessInst>(&I)) {
-      visitBeginAccess(BAI, accessSet);
+      visitBeginAccess(BAI);
       continue;
     }
     if (auto *EAI = dyn_cast<EndAccessInst>(&I)) {
-      visitEndAccess(EAI, accessSet);
+      visitEndAccess(EAI);
       continue;
     }
     if (auto fullApply = FullApplySite::isa(&I)) {
-      visitFullApply(fullApply, accessSet);
+      visitFullApply(fullApply);
     }
   }
-  LLVM_DEBUG(if (accessSet.hasConflictFreeAccess()) {
-    llvm::dbgs() << "Initializing no-conflict access out of bb"
-                 << BB->getDebugID() << "\n";
-    accessSet.dump();
-  });
-  if (BB->getTerminator()->isFunctionExiting())
-    assert(!accessSet.hasInScopeAccess() && "no postdominating end_access");
+}
 
-  // Initialize blockOutAccess for this block with the current access set.
-  accessSet.copyNoNestedConflictInto(blockOutAccess[BB]);
+void RegionStorageAnalysis::processLoopRegion(unsigned id) {
+  assert(regionToStorageMap.find(id) == regionToStorageMap.end() &&
+         "Should not process a region twice");
+  auto *region = LRFI->getRegion(id);
+  assert(!region->isBlock() && "Did not expect a block");
+  // A block's inital set is an empty access set
+  auto *storage = new RegionStorage(region, result);
+  regionToStorageMap.insert(std::make_pair(id, storage));
+  for (auto subID : region->getSubregions()) {
```

Comment:
Is this guaranteed RPO order? That's critical for the correctness of the data flow algorithm so needs to be explained.

Reply:
It is guaranteed RPO order. will explain that